### PR TITLE
Support new spawn

### PR DESCRIPTION
**Pull request by joii2020**
The simulator uses multithreading to simulate spawn, and each spawn will create a new thread.
After this change, need to use the matching [ckb-std](https://github.com/joii2020/ckb-std/tree/master) to use it in the CKB contract.

## Modify
* Support spawn (multithreading)
* Add testcase
* Add CI

### CODE REVIEW

**Review by XuJiandong — CHANGES_REQUESTED** [review]

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: src/spawn.rs [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        let out_fds = get_tx!(&TxContext::ctx_id())
+            .process(&Process::ctx_id())
+            .inherited_fds();
+        let len = out_fds.len().min(unsafe { *length });
+
+        Self::copy_fd(&out_fds[0..len], fds);
+        0
+    }
+
+    fn read(&self, fd: Fd, buf: *mut c_void, length: *mut usize) -> c_int {
```
Comment (XuJiandong) [inline src/spawn.rs:152]: If 100 bytes of data are written from one end of the pipe and the other end performs the following actions:
1. Reads 30 bytes
2. Reads 40 bytes
3. Reads 30 bytes

The simulator is unable to handle this correctly.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: src/global_data.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        Self(value)
+    }
+}
+impl From<TxID> for u64 {
+    fn from(value: TxID) -> Self {
+        value.0
+    }
+}
+impl TxID {
+    fn next(&mut self) -> Self {
```
Comment (mohanson) [inline src/global_data.rs:27]: Why do you modify the original data here, but clone a copy of the data as the return value at the same time?

Comment (joii2020) [reply src/global_data.rs:27]: In many places, the `next` step is `HashMap::insert`. This way, the return is more convenient to use.

File: .github/workflows/rust.yaml [hunk 3]
```diff
@@ -14,5 +14,13 @@ jobs:
 
     steps:
     - uses: actions/checkout@v3
+    - name: Install llvm 16
+      run: wget https://apt.llvm.org/llvm.sh && chmod +x llvm.sh && sudo ./llvm.sh 16 && rm llvm.sh
```
Comment (XuJiandong) [inline .github/workflows/rust.yaml:18]: use llvm-18

File: src/simulator_context.rs [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+        process.join_handle.take()
+    }
+
+    fn process_io(&mut self, fd: Option<&Fd>) {
+        // println!("==status 1: {:?}", self.process_status);
+
+        let mut update_rw: Vec<(usize, usize, bool)> = Vec::<(usize, usize, bool)>::new(); // Vec<(Read, Write)>
+        for i in 0..self.process_status.len() {
+            if let Some((_pid, rfd, rlen, _rbuf)) = self.process_status[i].read_wait() {
+                if rlen != &0 {
```
Comment (XuJiandong) [inline src/simulator_context.rs:264]: rlen can be zero. Any reading or writing zero bytes operation can block.

File: src/spawn.rs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+    unsafe { *({ pid }) = new_id.into() };
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn ckb_wait(pid: u64, code: *mut i8) -> c_int {
+    let pid: ProcID = pid.into();
+    if !get_cur_tx!().has_proc(&pid) {
+        return 5; // WaitFailure
```
Comment (XuJiandong) [inline src/spawn.rs:65]: define the errors first, no magic numbers.

**Review by XuJiandong — CHANGES_REQUESTED** [review]

File: src/simulator_context.rs (hunk 4 above)
Comment (joii2020) [reply src/simulator_context.rs:264]: This is just to refresh the read and write cache. The notification is executed in `notify_status`, and each status will be judged again.

File: src/lib.rs [hunk 6]
```diff
@@ -56,7 +66,7 @@ lazy_static! {
 }
 
 fn assert_vm_version() {
-    if SETUP.vm_version != 1 {
+    if SETUP.vm_version != 1 && SETUP.vm_version != 2 {
```
Comment (mohanson) [inline src/lib.rs:69]: There may be version 3 in the future, use `if SETUP.vm_version == 0` instead

**Review by XuJiandong — APPROVED** [review]

**Review by mohanson — APPROVED** [review]

**Pull request merged**